sources/niLang/src/System_Linux.cpp

```diff
@@ -809,25 +811,25 @@ class cLinuxWindow : public ni::ImplRC<ni::iOSWindow,ni::eImplFlags_Default,ni::
   }
 
   ///////////////////////////////////////////////
-  virtual tBool __stdcall UpdateWindow(tBool abBlockingMessages) niImpl {
+virtual tBool __stdcall UpdateWindow(tBool abBlockingMessages) niImpl {
     niCheckSilent(mHandle,eFalse);
 
-    // How many events are available in the queue.
-    dll_XSync(mpDisplay, False);
-    int events = dll_XEventsQueued(mpDisplay, QueuedAlready);
+    int events = dll_XEventsQueued(mpDisplay, QueuedAfterFlush);
     if (events > 0) {
-      // Limit amount of events we read at once
-      if (events > 5)
-        events = 5;
-      // Process events
-      while(events) {
+      // we stop processing events if it spends more than the average frame rate
+      const double maxProcessingTime = (1.0 / ni::GetLang()->GetAverageFrameRate()) * 1000.0;
+      double startTime = ni::GetLang()->TimerInSeconds() * 1000.0;
+      double currentTime = startTime;
```

Comment:
The Win32 implementation does this:
```
    if (mptrWindowThread.IsOK()) {
      if (abBlockingMessages && mptrWindowThread.IsOK()) {
        mptrWindowThread->WaitForMessage();
      }
    }
    else {
      // this way feels better, it make sure all input messages are processed in one shot...
      tU32 numProcessMessages = 0;
      if (abBlockingMessages) {
        WaitMessage();
      }
      MSG msg;
      while (niWin32API(PeekMessage)(&msg, NULL, 0, 0, PM_REMOVE)) {
        ::TranslateMessage(&msg);
        niWin32API(DispatchMessage)(&msg);
        ++numProcessMessages;
        if (numProcessMessages > 100)
          break;
      }
    }
```

The main thing being that it defaults to using a thread, otherwise it gives up after 100 messages per Update().

Ideally we'd update the events in a thread aswell I suppose, but that could mean rewriting this code substantially, for the moment I would settle for some fixed maximum number with a comment stating the potential issue and reasoning.

The timer is just bad it'll just lead to unpredictable behaviours.